### PR TITLE
bpo-44525: Specialize for calls to type and other builtin classes with 1 argument.

### PR DESCRIPTION
**Pull request by markshannon**
https://bugs.python.org/issue44525

### CODE REVIEW

**Review by Fidget-Spinner — COMMENTED** [review]
The code LGTM. Interested to see the benchmark results.

**Comment by markshannon** [comment]
[Benchmark results](https://gist.github.com/markshannon/8d581fd5663c0b107fabd7d91efa35a4) show a slight speedup, but very close to noise.

I think this is worth including because:
1. The nbody and spectral norm benchmarks show decent speedups and have `range(arg)` calls in the hot parts of the benchmark, suggesting that specialization of calls to `cls(one_arg)` is effective.
2. Some "real" code includes a lot of `type(obj)` calls (at least statically).

File: Python/ceval.c [hunk 1]
```diff
@@ -4854,6 +4854,41 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
             goto start_frame;
         }
 
+        TARGET(CALL_NO_KW_TYPE_1) {
+            assert(STACK_ADJUST_IS_RESET);
+            assert(GET_CACHE()->adaptive.original_oparg == 1);
+            PyObject *obj = TOP();
+            PyObject *callable = SECOND();
+            DEOPT_IF(callable != (PyObject *)&PyType_Type, CALL_NO_KW);
```
Comment (Fidget-Spinner) [inline Python/ceval.c:4862]: I don't remember if you like using the `Py_IS_TYPE` inline function, but just mentioning it here and leaving it up to you :).

File: Python/specialize.c [hunk 2]
```diff
@@ -1263,6 +1265,27 @@ specialize_class_call(
...
+    if (_Py_OPCODE(instr[-1]) == PRECALL_METHOD) {
+        SPECIALIZATION_FAIL(CALL_NO_KW, SPEC_FAIL_METHOD_CALL_CLASS);
+        return -1;
+    }
+    if (tp->tp_new == PyBaseObject_Type.tp_new) {
+        SPECIALIZATION_FAIL(CALL_NO_KW, SPEC_FAIL_PYTHON_CLASS);
+        return -1;
+    }
+    if (nargs == 1) {
+        if (tp == &PyType_Type) {
```
Comment (Fidget-Spinner) [inline Python/specialize.c:1279]: Ditto

**Review by Fidget-Spinner — APPROVED** [review]
LGTM.

File: Python/ceval.c (hunk 1 above)
Comment (markshannon) [reply Python/ceval.c:4862]: I'll skip on the obfuscation macro :slightly_smiling_face:

**Pull request merged**